src/ERC3009.sol

```diff
@@ -0,0 +1,184 @@
+// SPDX-License-Identifier: GPL-3.0
+
+pragma solidity 0.8.23;
+
+import { IERC3009 } from "./interfaces/IERC3009.sol";
+import { ERC712 } from "./libs/ERC712.sol";
+
+import { StatefulERC712 } from "./StatefulERC712.sol";
+
+abstract contract ERC3009 is IERC3009, StatefulERC712 {
+    // keccak256("TransferWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
+    bytes32 public constant TRANSFER_WITH_AUTHORIZATION_TYPEHASH =
+        0x7c7c6cdb67a18743f49ec6fa9b35f50d52ed05cbed4cc592e13b44501c1a2267;
+
+    // keccak256("ReceiveWithAuthorization(address from,address to,uint256 value,uint256 validAfter,uint256 validBefore,bytes32 nonce)")
+    bytes32 public constant RECEIVE_WITH_AUTHORIZATION_TYPEHASH =
+        0xd099cc98ef71107a616c4f0f941f04c322d8e254fe26b3c6668db87aae413de8;
+
+    // keccak256("CancelAuthorization(address authorizer,bytes32 nonce)")
+    bytes32 public constant CANCEL_AUTHORIZATION_TYPEHASH =
+        0x158b0a9edf7a828aad02f63cd515c68ef2f50ba807396f6d12842833a1597429;
+
+    /// @dev authorizer => nonce => state (true = used / false = unused)
+    mapping(address authorizer => mapping(bytes32 nonce => bool isNonceUsed)) internal _authorizationStates;
```

Comment:
This can be public if `function authorizationState` is not virtual. I probably set a bad precedent with `_nonces` in `StatefulERC712`.

Reply:
`authorizationState` is external si we can keep this mapping internal.
Why would we need the function to be `virtual`?

Reply:
Fixed in this commit: https://github.com/MZero-Labs/common/pull/7/commits/c98a486dcf668a367f03459dc7be7114682a9244